engine/engine.go

```diff
@@ -556,49 +569,116 @@ func (e *Engine) publishAdvForIndex(ctx context.Context, contextID []byte, md me
+	defaultProviderID := e.provider.ID
+	if provider == "" {
+		provider = defaultProviderID
+	}
+	if provider == defaultProviderID {
+		return datastore.NewKey(keyToCidMapPrefix + string(contextID))
+	} else {
+		return datastore.NewKey(keyToCidMapPrefix + string(contextID) + "/" + provider.String())
+	}
+}
+
+func (e *Engine) cidToKeyKey(c cid.Cid) datastore.Key {
+	return datastore.NewKey(cidToKeyMapPrefix + c.String())
+}
+
+func (e *Engine) cidToProviderAndKeyKey(c cid.Cid) datastore.Key {
+	return datastore.NewKey(cidToProviderAndKeyMapPrefix + c.String())
+}
+
+func (e *Engine) keyToMetadataKey(provider peer.ID, contextID []byte) datastore.Key {
+	defaultProviderID := e.provider.ID
+	if provider == "" {
+		provider = defaultProviderID
+	}
+	if provider == defaultProviderID {
+		return datastore.NewKey(keyToMetadataMapPrefix + string(contextID))
+	} else {
+		return datastore.NewKey(keyToMetadataMapPrefix + string(contextID) + "/" + provider.String())
+	}
+}
+
+func (e *Engine) putKeyCidMap(ctx context.Context, provider peer.ID, contextID []byte, c cid.Cid) error {
 	// Store the map Key-Cid to know what CidLink to put in advertisement when
 	// notifying about a removal.
-	err := e.ds.Put(ctx, datastore.NewKey(keyToCidMapPrefix+string(contextID)), c.Bytes())
+
+	err := e.ds.Put(ctx, e.keyToCidKey(provider, contextID), c.Bytes())
 	if err != nil {
 		return err
 	}
 	// And the other way around when graphsync is making a request, so the
 	// lister in the linksystem knows to what contextID the CID referrs to.
-	return e.ds.Put(ctx, datastore.NewKey(cidToKeyMapPrefix+c.String()), contextID)
+	// it's enough for us to store just a single mapping of cid to provider and context to generate chunks
+	var buffer bytes.Buffer
+	err = gob.NewEncoder(&buffer).Encode(&providerAndContext{Provider: provider, ContextID: contextID})
+	if err != nil {
+		return err
+	}
+	return e.ds.Put(ctx, e.cidToProviderAndKeyKey(c), buffer.Bytes())
 }
 
-func (e *Engine) getKeyCidMap(ctx context.Context, contextID []byte) (cid.Cid, error) {
-	b, err := e.ds.Get(ctx, datastore.NewKey(keyToCidMapPrefix+string(contextID)))
+func (e *Engine) getKeyCidMap(ctx context.Context, provider peer.ID, contextID []byte) (cid.Cid, error) {
+	b, err := e.ds.Get(ctx, e.keyToCidKey(provider, contextID))
 	if err != nil {
 		return cid.Undef, err
 	}
 	_, d, err := cid.CidFromBytes(b)
 	return d, err
 }
 
-func (e *Engine) deleteKeyCidMap(ctx context.Context, contextID []byte) error {
-	return e.ds.Delete(ctx, datastore.NewKey(keyToCidMapPrefix+string(contextID)))
+func (e *Engine) deleteKeyCidMap(ctx context.Context, provider peer.ID, contextID []byte) error {
+	return e.ds.Delete(ctx, e.keyToCidKey(provider, contextID))
 }
 
-func (e *Engine) deleteCidKeyMap(ctx context.Context, c cid.Cid) error {
-	return e.ds.Delete(ctx, datastore.NewKey(cidToKeyMapPrefix+c.String()))
+func (e *Engine) deleteCidKeyMap(ctx context.Context, provider peer.ID, c cid.Cid) error {
```

Comment:
this is because neither cidToKey nor cidToProvider indexes have `providerId` in them. The former is a legacy index that didn't have a support for multiple providers. The latter stores mapping for *any* provider (whoever is the latest) as it's good enough to generate chunks when needed. I removed the extra parameter.

Reply:
Remove provider parameter. It's not used because it's not required for legacy index as there was only one provider neither it's required for the  new index as it stores *any* `cid -> (context, provider)` mapping as any is good enough to regenerate chunks